examples/nine-pt-stencil-solver/nine-pt-stencil-solver.cpp

```diff
@@ -284,48 +290,52 @@ int main(int argc, char *argv[])
                   << std::endl;
         std::exit(-1);
     }
+    using ValueType = double;
+    using IndexType = int;
 
     const int discretization_points = argc >= 2 ? std::atoi(argv[1]) : 100;
     const auto executor_string = argc >= 3 ? argv[2] : "reference";
-    const double alpha_c = argc >= 4 ? std::atof(argv[3]) : default_alpha;
-    const double beta_c = argc >= 5 ? std::atof(argv[4]) : default_beta;
-    const double gamma_c = argc >= 6 ? std::atof(argv[5]) : default_gamma;
+    const ValueType alpha_c = argc >= 4 ? std::atof(argv[3]) : default_alpha;
+    const ValueType beta_c = argc >= 5 ? std::atof(argv[4]) : default_beta;
+    const ValueType gamma_c = argc >= 6 ? std::atof(argv[5]) : default_gamma;
 
     // clang-format off
-    std::array<double, 9> coefs{
+    std::array<ValueType, 9> coefs{
         gamma_c, beta_c, gamma_c,
-	beta_c, alpha_c, beta_c,
+	      beta_c, alpha_c, beta_c,
         gamma_c, beta_c, gamma_c};
     // clang-format on
 
     const auto dp = discretization_points;
     const size_t dp_2 = dp * dp;
 
     // problem:
-    auto correct_u = [](double x, double y) { return x * x * x + y * y * y; };
-    auto f = [](double x, double y) { return 6 * x + 6 * y; };
+    auto correct_u = [](ValueType x, ValueType y) {
+        return x * x * x + y * y * y;
+    };
+    auto f = [](ValueType x, ValueType y) { return 6 * x + 6 * y; };
 
     // matrix
-    std::vector<int> row_ptrs(dp_2 + 1);
-    std::vector<int> col_idxs((3 * dp - 2) * (3 * dp - 2));
-    std::vector<double> values((3 * dp - 2) * (3 * dp - 2));
+    std::vector<IndexType> row_ptrs(dp_2 + 1);
+    std::vector<IndexType> col_idxs((3 * dp - 2) * (3 * dp - 2));
+    std::vector<ValueType> values((3 * dp - 2) * (3 * dp - 2));
     // right hand side
-    std::vector<double> rhs(dp_2);
+    std::vector<ValueType> rhs(dp_2);
     // solution
-    std::vector<double> u(dp_2, 0.0);
+    std::vector<ValueType> u(dp_2, 0.0);
 
     generate_stencil_matrix(dp, row_ptrs.data(), col_idxs.data(), values.data(),
                             coefs.data());
     // looking for solution u = x^3: f = 6x, u(0) = 0, u(1) = 1
     generate_rhs(dp, f, correct_u, rhs.data(), coefs.data());
 
-    auto start_time = std::chrono::steady_clock::now();
+    ValueType reduction_factor = 1e-7;
 
+    auto start_time = std::chrono::steady_clock::now();
     solve_system(executor_string, dp, row_ptrs.data(), col_idxs.data(),
-                 values.data(), rhs.data(), u.data(), 1e-12);
-
+                 values.data(), rhs.data(), u.data(), reduction_factor);
     auto stop_time = std::chrono::steady_clock::now();
-    double runtime_duration =
+    ValueType runtime_duration =
```

Comment:
nit: The time type should not be impacted by `ValueType`.
```suggestion
    auto runtime_duration =
```